extensions/amp-viewer-integration/0.1/keyboard-handler.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {KeyCodes} from '../../../src/utils/key-codes';
+import {dict} from '../../../src/utils/object';
+import {listen} from '../../../src/event-helper';
+
+/**
+ * The list of keyboard event properites to forward to the viewer. This should
+ * be kept up-to-date with
+ * https://www.w3.org/TR/uievents/#keyboardevent-keyboardevent.
+ *
+ * @type {!Array<string>}
+ */
+const eventProperties = [
+  'key',
+  'code',
+  'location',
+
+  'ctrlKey',
+  'shiftKey',
+  'altKey',
+  'metaKey',
+
+  'repeat',
+  'isComposing',
+
+  // Properties for legacy user agents.
+  'charCode',
+  'keyCode',
+  'which',
+];
+
+/**
+ * Forwards keyboard events from the AMP doc to the viewer in the format of a
+ * `KeyboardEventInit` object (http://mdn.io/KeyboardEvent/KeyboardEvent).
+ * `JsonObject`.
+ *
+ * Keyboard events that are forwarded must meet the *one* of the following
+ * conditions:
+ *
+ * - The key is the escape key
+ * - The focus is on a checkbox and the key is not the space key
+ * - The focus is not on any input control, including elements with the
+ *   `contenteditable` attribute
+ *
+ * @package @final
+ */
+export class KeyboardHandler {
+  /**
+   * @param {!Window} win
+   * @param {!./messaging/messaging.Messaging} messaging
+   */
+  constructor(win, messaging) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @const @private {!./messaging/messaging.Messaging} */
+    this.messaging_ = messaging;
+
+    this.listenForKeyboardEvents_();
+  }
+
+  /** @private */
+  listenForKeyboardEvents_() {
+    const handleEvent = this.handleEvent_.bind(this);
+    listen(this.win, 'keydown', handleEvent);
+    listen(this.win, 'keypress', handleEvent);
+    listen(this.win, 'keyup', handleEvent);
```

Comment:
Curious why we need to forward `keypress` and `keyup`?

Reply:
For `keypress`, some viewer may insist on using `keypress` instead of `keydown` for shortcuts that produce a character. For `keyup`, the viewer may want to prevent repeatedly triggering something when the keyboard shortcut is kept depressed.

Reply:
For example, if the application on the viewer side uses the Closure event library to handle keyboard shortcuts, then it will need `keypress` because the library apparently only handles the event upon `keypress` for keys that produce printable characters:

https://github.com/google/closure-library/blob/db35cf1524b36a50d021fb6cf47271687cc2ea33/closure/goog/events/keyhandler.js#L328